From b765a34bd7dc50ecf0f9d65256321b79ed4599ce Mon Sep 17 00:00:00 2001
From: Roy.Li <rongqing.li@windriver.com>
Date: Tue, 30 Aug 2011 09:17:41 +0800
Subject: [PATCH] gianfar: satisfy maxfrm hardware limitation by reducing mtu

When mtu is 9600, maxfrm will be 9728(512 x 19). Because the maxfrm is
aligned to 512 bytes as defined INCREMENTAL_BUFFER_SIZE in gianfar.h.

But the MPC83xx Reference Manual says the maxfrm can not be greater
than 0x2580(9600 bytes).

So this fix is doing the below change.

	1) Limit the mtu range to 68 <= new_mtu <= 9000 bytes.
	2) Then add the necessary additional hw headers/padding,
	 and align to 512 bytes.
	3) Check whether the resulting size does exceed 9600
	(hw limitation) or not, if yes, set it to 9600.

Fixed-by: Borman, David <david.borman@windriver.com>
Integrated-by: Roy Li <rongqing.li@windriver.com>
---
 drivers/net/gianfar.c |   11 +++++++----
 drivers/net/gianfar.h |    4 ++--
 2 files changed, 9 insertions(+), 6 deletions(-)

diff --git a/drivers/net/gianfar.c b/drivers/net/gianfar.c
index 959d4ae..8ee8d62 100644
--- a/drivers/net/gianfar.c
+++ b/drivers/net/gianfar.c
@@ -3246,16 +3246,16 @@ static int gfar_change_mtu(struct net_device *dev, int new_mtu)
 	int oldsize = priv->rx_buffer_size;
 	int frame_size = new_mtu + ETH_HLEN;
 
-	if (priv->vlgrp)
-		frame_size += VLAN_HLEN;
-
-	if ((frame_size < 64) || (frame_size > JUMBO_FRAME_SIZE)) {
+	if ((new_mtu < 68) || (new_mtu > JUMBO_FRAME_SIZE)) {
 		if (netif_msg_drv(priv))
 			printk(KERN_ERR "%s: Invalid MTU setting\n",
 					dev->name);
 		return -EINVAL;
 	}
 
+	if (priv->vlgrp)
+		frame_size += VLAN_HLEN;
+
 	if (gfar_uses_fcb(priv))
 		frame_size += GMAC_FCB_LEN;
 
@@ -3265,6 +3265,9 @@ static int gfar_change_mtu(struct net_device *dev, int new_mtu)
 	    (frame_size & ~(INCREMENTAL_BUFFER_SIZE - 1)) +
 	    INCREMENTAL_BUFFER_SIZE;
 
+	if (tempsize > JUMBO_BUFFER_SIZE)
+		tempsize = JUMBO_BUFFER_SIZE;
+
 	/* Only stop and start the controller if it isn't already
 	 * stopped, and we changed something */
 	if ((oldsize != tempsize) && (dev->flags & IFF_UP))
diff --git a/drivers/net/gianfar.h b/drivers/net/gianfar.h
index 1850be5..c766eae 100644
--- a/drivers/net/gianfar.h
+++ b/drivers/net/gianfar.h
@@ -131,8 +131,8 @@ static inline int devfp_register_tx_hook(devfp_hook_t hook)
 #define DEFAULT_WK_BUFFER_SIZE	2048
 #define TX_RING_MOD_MASK(size) (size-1)
 #define RX_RING_MOD_MASK(size) (size-1)
-#define JUMBO_BUFFER_SIZE 9728
-#define JUMBO_FRAME_SIZE 9600
+#define JUMBO_BUFFER_SIZE 9600
+#define JUMBO_FRAME_SIZE 9000
 
 #define DEFAULT_FIFO_TX_THR 0x100
 #define DEFAULT_FIFO_TX_STARVE 0x40
-- 
1.7.0.4

